gammapy/cube/exposure.py

```diff
@@ -42,3 +45,59 @@ def make_map_exposure_true_energy(pointing, livetime, aeff, geom):
     exposure = (exposure * livetime).to('m2 s')
 
     return WcsNDMap(geom, exposure.value, unit=exposure.unit)
+
+def make_map_exposure_reco_energy(exposure_map, spectrum=None, edisp=None):
+    """Create an exposure map in reco energy from an exposure map in true energy.
+
+    Exposure in true energy is weighted with an input spectrum and redistributed in
+    reco energy with the input energy dispersion.
+
+    Parameters
+    ----------
+    exposure_map : `~gammapy.maps.Map`
+        Input exposure map in true energy. unit should have dimension of m2s
+    spectrum : `~gammapy.spectrum.models.SpectralModel`, default is None
+        Spectral model to use to weight exposure in true energy
+        If None is passed, a power law of photon index -2 is assumed.
+    edisp : `~gammapy.irf.EnergyDispersion`, default is None.
+        Energy Dispersion response to redistribute true energies to reco energies
+        If no energy dispersion is passed, a diagonal one is assumed with identical true and reco
+        energy bins. This is the default behaviour.
+
+    Returns
+    -------
+    exposure_image : `~gammapy.maps.Map`
+        Resulting image in reco energy. The unit is the same as the input map.
+    """
+    expo_map = exposure_map.copy()
+    etrue_axis = expo_map.geom.get_axis_by_name("energy")
+    etrue_center = etrue_axis.center * etrue_axis.unit
+    etrue_edges = etrue_axis.edges * etrue_axis.unit
+    binsize = np.diff(etrue_edges)
+
+    if spectrum is None:
+        spectrum = PowerLaw(index=2.0)
+    weights = spectrum(etrue_center)*binsize
+    weights /= weights.sum()
+
+    for img, idx in expo_map.iter_by_image():
+        img *= weights[idx].value
+
+    if edisp is None:
+        edisp = EnergyDispersion.from_diagonal_matrix(etrue_edges, etrue_edges)
+
+    # Here we do the convolution with the edisp matrix
+    a = np.rollaxis(expo_map.data, 0, 3)
+    b = np.dot(a, edisp.pdf_matrix)
```

Comment:
We shouldn't duplicate this np.rollaxis, np.dot, np.rollaxis stuff, no?
Already figured out and tested here:
http://docs.gammapy.org/dev/_modules/gammapy/cube/fit.html#MapEvaluator.apply_edisp

Of course, we could also call from MapEvaluator into a function, but it shouldn't be duplicated IMO.

Reply:
We will have other usages of an `apply_edisp` taking a `Map` and an `EnergyDispersion` as input; for instance, if we want a PSFKernel in reco_energy for morphological analysis.

Therefore, I propose to create a stand-alone function `apply_edisp(map, edisp)`. It can be based on a private function `_map_dot_product(array, array)` that will just to the axis rotation and dot product on the data. The latter can be also used in `MapEvaluator.apply_edisp()`.

Is this OK @cdeil, @adonath ?

Reply:
I don't have the big picture yet how this function and the MapEvaluator relate.

This is on the line between MapMaker and MapEvaluator, could go either way.

This could also become a method on MapEvaluator, because it's doing Spectrum / npred related computations. Note that you can have `@staticmethod` just located on MapEvalutor, to avoid 10 standalone functions.

If you want to pursue this as a standalone function, you could also just make `MapEvaluator.apply_edisp` a `@staticmethod and call it here from now. Then the code duplication with the rollaxis is avoided. Or move it to a separate function and call it from both places. But I'm not super excited about ending up with ~ 10 functions that each just have 3 lines, like the ones we now have on the MapEvaluator.

@registerrier - ?